File: Python/Intersections.py
#Author-Autodesk Inc.
#Description-Caculate the intersections between the selected curve/surface/body/component/occurrence and curve/surface.
# non planar surface does not support for now

import adsk.core, adsk.fusion, traceback
import os

pi = 3.1415926
nearZero = 0.000001
# global set of event handlers to keep them referenced for the duration of the command
handlers = []
app = adsk.core.Application.get()
if app:
    ui = app.userInterface

class IntersectionCommandExecuteHandler(adsk.core.CommandEventHandler):
    def __init__(self):
        super().__init__()
    def notify(self, args):
        try:
            command = args.firingEvent.sender
            inputs = command.commandInputs

            input0 = inputs[0];
            sel0 = input0.selection(0);

            input1 = inputs[1];
            sel1 = input1.selection(0);

            intersections = Intersections();
            intersections.Execute(sel0.entity, sel1.entity);
        except:
            if ui:
                ui.messageBox('Failed:\n{}'.format(traceback.format_exc()))

class IntersectionCommandDestroyHandler(adsk.core.CommandEventHandler):
    def __init__(self):
        super().__init__()
    def notify(self, args):
        try:
            # when the command is done, terminate the script
            # this will release all globals which will remove all event handlers
            adsk.terminate()
        except:
            if ui:
                ui.messageBox('Failed:\n{}'.format(traceback.format_exc()))

class IntersectionValidateInputHandler(adsk.core.ValidateInputsEventHandler):
    def __init__(self):
        super().__init__()
       
    def notify(self, args):
        try:
            sels = ui.activeSelections;
            if len(sels) == 2:
                args.areInputsValid = True
            else:
                args.areInputsValid = False
        except:
            if ui:
                ui.messageBox('Failed:\n{}'.format(traceback.format_exc()))

class IntersectionCommandCreatedHandler(adsk.core.CommandCreatedEventHandler):
    def __init__(self):
        super().__init__()
    def notify(self, args):
        try:
            cmd = args.command
            onExecute = IntersectionCommandExecuteHandler()
            cmd.execute.add(onExecute)
            onDestroy = IntersectionCommandDestroyHandler()
            cmd.destroy.add(onDestroy)

            onValidateInput = IntersectionValidateInputHandler()
            cmd.validateInputs.add(onValidateInput)
            # keep the handler referenced beyond this function
            handlers.append(onExecute)
            handlers.append(onDestroy)
            handlers.append(onValidateInput)
            #define the inputs
            inputs = cmd.commandInputs
            i1 = inputs.addSelectionInput('entity', 'Entity One', 'Please select a curve, planear entity or a BRepBody, Component, Occurrence')

            i1.addSelectionFilter(adsk.core.SelectionCommandInput.Edges);
            i1.addSelectionFilter(adsk.core.SelectionCommandInput.PlanarFaces);
            i1.addSelectionFilter(adsk.core.SelectionCommandInput.SketchCurves);
            i1.addSelectionFilter(adsk.core.SelectionCommandInput.ConstructionLines);
            i1.addSelectionFilter(adsk.core.SelectionCommandInput.ConstructionPlanes);
            i1.addSelectionFilter(adsk.core.SelectionCommandInput.Bodies);
            i1.addSelectionFilter(adsk.core.SelectionCommandInput.Occurrences);
            i1.addSelectionFilter(adsk.core.SelectionCommandInput.RootComponents);

            i2 = inputs.addSelectionInput('sectionentity', 'Entity Two', 'Please select a linear or planear entity')

            i2.addSelectionFilter(adsk.core.SelectionCommandInput.PlanarFaces);
            i2.addSelectionFilter(adsk.core.SelectionCommandInput.LinearEdges);
            i2.addSelectionFilter(adsk.core.SelectionCommandInput.SketchLines);
            i2.addSelectionFilter(adsk.core.SelectionCommandInput.ConstructionLines);
            i2.addSelectionFilter(adsk.core.SelectionCommandInput.ConstructionPlanes);
        except:
            if ui:
                ui.messageBox('Failed:\n{}'.format(traceback.format_exc()))

class Intersections:
    def Execute(self, entityOne, entityTwo):
        #caculate the intersections
        sectionResults = adsk.core.ObjectCollection.create()
        def getGeometry(entity):
                geom = entity
                if isinstance(entity, adsk.fusion.BRepFace) or \
                   isinstance(entity, adsk.fusion.BRepEdge) or \
                   isinstance(entity, adsk.fusion.ConstructionAxis) or\
                   isinstance(entity,adsk.fusion.ConstructionPlane):
                    geom = entity.geometry
                elif isinstance(entity, adsk.fusion.SketchCurve):
                    geom = entity.worldGeometry;
                return geom;

        def intersectWith(surfaceOrCurve, section):
            surfaceOrCurve = getGeometry(surfaceOrCurve)

            section = getGeometry(section)
            result = None
            if isinstance(surfaceOrCurve, adsk.core.Curve3D):
                result = section.intersectWithCurve(surfaceOrCurve)
            else:
                if surfaceOrCurve.surfaceType == adsk.core.SurfaceTypes.PlaneSurfaceType and section.surfaceType == adsk.core.SurfaceTypes.PlaneSurfaceType :
                    result = section.intersectWithPlane(surfaceOrCurve)
                    if result:
                        sectionResults.add(result)
                    return 
            if result:
                for resultI in result:
                    sectionResults.add(resultI)

        def intersectWithBody(body, section):
            fs = body.faces;
            for fsI in fs:
                intersectWith(fsI, section)

        def intersectWithComponent(comp, occ, section):
            if isinstance(comp,adsk.fusion.Component):
                bodies = comp.bRepBodies
                for body in bodies:
                    if(not body):
                        continue
                    if occ :
                        body = body.createForAssemblyContext(occ)

                    intersectWithBody(body, section)

            childOccs = None
            if occ :
                childOccs = occ.childOccurrences
            else:
                childOccs = comp.occurrences
            
            for childOcc in childOccs:
                if not childOcc:
                    continue
                intersectWithComponent(childOcc.component, childOcc, section);

        if isinstance(entityOne,adsk.fusion.Component):
            intersectWithComponent(entityOne, None, entityTwo)

        elif isinstance(entityOne,adsk.fusion.Occurrence):
            intersectWithComponent(entityOne.component, entityOne, entityTwo)

        elif isinstance(entityOne, adsk.fusion.BRepBody):
            intersectWithBody(entityOne, entityTwo)

        else:
            intersectWith(entityOne, entityTwo)

        if len(sectionResults) == 0:
            if ui:
                ui.messageBox('No intersection found')
            return

        def isPlanearEntity(entity):
            planearEnt = False
            if isinstance(entity, adsk.fusion.ConstructionPlane):
                planearEnt = True
            elif isinstance(entity, adsk.fusion.BRepFace):
                sur = entity.geometry

                if(sur.surfaceType == adsk.core.SurfaceTypes.PlaneSurfaceType):
                    planearEnt = True

            return planearEnt

        doc = app.activeDocument
        d = doc.design
        rootComp = d.rootComponent

        sketch = None
        if isPlanearEntity(entityTwo):
            sketch = rootComp.sketches.add(entityTwo)

        for geom in sectionResults:
            if not geom:
                continue

            if isinstance(geom,adsk.core.Point3D):
                ptInput = rootComp.constructionPoints.createInput()
                ptInput.setByPoint(geom)
                rootComp.constructionPoints.add(ptInput)
            elif isinstance(geom,adsk.core.Curve3D) and sketch:
                m = sketch.transform;
                m.invert();
                geom.transformBy(m);
                sketchCurve = None
                if isinstance(geom, adsk.core.Line3D):
                    sketchCurve = sketch.sketchCurves.sketchLines.addByTwoPoints(geom.startPoint, geom.endPoint);

                elif isinstance(geom,adsk.core.Arc3D):
                    sweepAngle = 2 * pi if abs(geom.endAngle - geom.startAngle) < nearZero else geom.startAngle
                    sketchCurve = sketch.sketchCurves.sketchArcs.addByCenterStartSweep(geom.center, geom.startPoint, sweepAngle)

                elif isinstance(geom,adsk.core.Circle3D):
                    sketchCurve = sketch.sketchCurves.sketchCircles.addByCenterRadius(geom.center, geom.radius)

                elif isinstance (geom,adsk.core.Ellipse3D):
                    curveEva = geom.evaluator

                    startParameter = None
                    endParameter = None
                    curveEva.getParameterExtents(startParameter, endParameter)

                    pointOnCurve = None
                    curveEva.getPointAtParameter((startParameter + endParameter)/3, pointOnCurve)

                    majorAxisPoint = geom.center
                    majorAxisVec = geom.majorAxis

                    majorAxisVec.scaleBy(geom.majorRadius)
                    majorAxisPoint.translateBy(majorAxisVec)

                    sketchCurve = sketch.sketchCurves.sketchEllipses.add(geom.center, majorAxisPoint, pointOnCurve)

                elif isinstance(geom, adsk.core.NurbsCurve3D):
                    pts = geom.controlPoints

                    ptCol = adsk.core.ObjectCollection.create()

                    for ptsI in pts:
                        ptCol.add(ptsI)

                    sketchCurve = sketch.sketchCurves.SketchFittedSplines.add(ptCol)

                elif isinstance(geom,adsk.core.InfiniteLine3D):
                    start = geom.origin
                    end = geom.origin
                    dir = geom.direction
                    dir.scaleBy(10)
                    end.translateBy(dir)
                    sketchCurve = sketch.sketchCurves.sketchLines.addByTwoPoints(start, end)

                if sketchCurve:
                    sketchCurve.isConstruction = True

def main():
    try:
        commandDefinitions = ui.commandDefinitions
        # check the command exists or not
        cmdDef = commandDefinitions.itemById('IntersectionCMDDef')
        if not cmdDef:
            resourceDir = os.path.join(os.path.dirname(os.path.realpath(__file__)), 'resources') # absolute resource file path is specified
            cmdDef = commandDefinitions.addButtonDefinition('IntersectionCMDDef',
                    'Intersections',
                    'Calculate the intersections of two selected entities',
                    resourceDir)

        onCommandCreated = IntersectionCommandCreatedHandler()
        cmdDef.commandCreated.add(onCommandCreated)
        # keep the handler referenced beyond this function
        handlers.append(onCommandCreated)
        inputs = adsk.core.NamedValues.create()
        cmdDef.execute(inputs)

        # prevent this module from being terminate when the script returns, because we are waiting for event handlers to fire
        adsk.autoTerminate(False)
    except:
        if ui:
            ui.messageBox('Failed:\n{}'.format(traceback.format_exc()))

main()

File: JavaScript/Intersections.js
//Author-Autodesk Inc.
//Description-Caculate the intersections between the selected curve/surface/body/component/occurrence and curve/surface.
// non planar surface does not support for now
/*globals adsk*/
(function () {

    "use strict";
    
    var app = adsk.core.Application.get(), ui;
    if (app) {
        ui = app.userInterface;
    }
    
    var pi = 3.1415926;
    var nearZero = 0.000001;
    var Intersections = function(entityOne, entityTwo) {
        
        this.Execute =  function () {
            
            //caculate the intersections
            var sectionResults = adsk.core.ObjectCollection.create();
            
            function getGeometry(entity) {
                var geom = entity;
                if(entity instanceof adsk.fusion.BRepFace ||
                   entity instanceof adsk.fusion.BRepEdge ||
                   entity instanceof adsk.fusion.ConstructionAxis ||
                   entity instanceof adsk.fusion.ConstructionPlane) {
                    geom = entity.geometry;
                }
                else if(entity instanceof adsk.fusion.SketchCurve) {
                    geom = entity.worldGeometry;
                }
                
                return geom;
            }
            
            function intersectWith(surfaceOrCurve, section) {
                
                surfaceOrCurve = getGeometry(surfaceOrCurve);
                
                section = getGeometry(section);
                
                var result = null;
                if(surfaceOrCurve instanceof adsk.core.Curve3D) {
                    result = section.intersectWithCurve(surfaceOrCurve);
                }
                else {
                    if(surfaceOrCurve.surfaceType == adsk.core.SurfaceTypes.PlaneSurfaceType && 
                      section.surfaceType == adsk.core.SurfaceTypes.PlaneSurfaceType) {
                        result = section.intersectWithPlane(surfaceOrCurve);
                        if(result) {
                            sectionResults.add(result);
                        }
                        return;
                    }
                }
                
                if(result) {
                    for(var i = 0; i < result.count; i++) {
                        sectionResults.add(result.item(i));
                    }
                }
            }
            
            function intersectWithBody(body, section) {
                
                var fs = body.faces;
                
                for(var i = 0; i < fs.count; i++) {
                    intersectWith(fs.item(i), section);
                }
            }
            
            function intersectWithComponent(comp, occ, section) {
                
                if(comp instanceof adsk.fusion.Component) {
                    var bodies = comp.bRepBodies;
                    
                    for(var i = 0; i < bodies.count; i++) {
                        var body = bodies.item(i);
                        if(!body) {
                            continue;
                        }
                        
                        if(occ) {
                            body = body.createForAssemblyContext(occ);
                        }
                        intersectWithBody(body, section);
                    }
                }
                
                var childOccs = null;
                if(occ) {
                    childOccs = occ.childOccurrences;
                }
                else {
                    childOccs = comp.occurrences;
                }
                
                for(var n = 0; n < childOccs.count; n++) {
                    var childOcc = childOccs.item(n);
                    if(!childOcc) {
                        continue;
                    }
                    
                    intersectWithComponent(childOcc.component, childOcc, section);
                }
            }
            
            if(entityOne instanceof adsk.fusion.Component) {
                intersectWithComponent(entityOne, null, entityTwo);
            }
            else if(entityOne instanceof adsk.fusion.Occurrence) {
                intersectWithComponent(entityOne.component, entityOne, entityTwo);
            }
            else if(entityOne instanceof adsk.fusion.BRepBody) {
                intersectWithBody(entityOne, entityTwo);
            }
            else {
                intersectWith(entityOne, entityTwo);
            }
            
            var app = adsk.core.Application.get();
            
            if(sectionResults.count === 0) {
                var ui = app.userInterface;
                if (ui) {
                    ui.messageBox('No intersection found');
                }
                
                return;
            }
            
            // print the results
            function isPlanearEntity(entity) {
                
                var planearEnt = false;
                if(entity instanceof adsk.fusion.ConstructionPlane) {
                    planearEnt = true;
                }
                else if(entity instanceof adsk.fusion.BRepFace) {
                    var sur = entity.geometry;
                    
                    if(sur.surfaceType == adsk.core.SurfaceTypes.PlaneSurfaceType) {
                        planearEnt = true;
                    }
                }
                
                return planearEnt;
            }
            
            var doc = app.activeDocument;
            
            var d = doc.design;
            
            var rootComp = d.rootComponent;
            
            var sketch = null;
            if(isPlanearEntity(entityTwo)) {
                sketch = rootComp.sketches.add(entityTwo);
            }
            
            for(var i = 0; i < sectionResults.count; i++) {
                var geom = sectionResults.item(i);
                if(!geom) {
                    continue;
                }
                
                if(geom instanceof adsk.core.Point3D) {
                    var ptInput = rootComp.constructionPoints.createInput();
                    
                    ptInput.setByPoint(geom);
                    rootComp.constructionPoints.add(ptInput);
                }
                else if(geom instanceof adsk.core.Curve3D && sketch) {
                    var m = sketch.transform;
                    m.invert();
                    geom.transformBy(m);
                    var sketchCurve;
                    if(geom instanceof adsk.core.Line3D) {
                        sketchCurve = sketch.sketchCurves.sketchLines.addByTwoPoints(geom.startPoint, geom.endPoint);
                    }
                    else if(geom instanceof adsk.core.Arc3D) {
                        var sweepAngle = Math.abs(geom.endAngle - geom.startAngle) < nearZero ? 2 * pi : geom.endAngle - geom.startAngle;
                        sketchCurve = sketch.sketchCurves.sketchArcs.addByCenterStartSweep(geom.center, geom.startPoint, sweepAngle);
                    }
                    else if(geom instanceof adsk.core.Circle3D) {
                        sketchCurve = sketch.sketchCurves.sketchCircles.addByCenterRadius(geom.center, geom.radius);
                    }
                    else if(geom instanceof adsk.core.Ellipse3D) {
                        var curveEva = geom.evaluator;
                        
                        var startParameter, endParameter;
                        var bOk = curveEva.getParameterExtents(startParameter, endParameter);
                        
                        var pointOnCurve = null;
                        bOk = curveEva.getPointAtParameter((startParameter + endParameter)/3, pointOnCurve);
                        
                        var majorAxisPoint = geom.center;
                        
                        var majorAxisVec = geom.majorAxis;

                        bOk = majorAxisVec.scaleBy(geom.majorRadius);
                        
                        bOk = majorAxisPoint.translateBy(majorAxisVec);
                        
                        sketchCurve = sketch.sketchCurves.sketchEllipses.add(geom.center, majorAxisPoint, pointOnCurve);
                    }
                    else if(geom instanceof adsk.core.NurbsCurve3D) {
                        var pts = geom.controlPoints;
                        var ptCount = geom.controlPointCount;
                        
                        var ptCol = adsk.core.ObjectCollection.create();
                        
                        for(var j = 0; j < ptCount; j++) {
                            ptCol.add(pts[i]);
                        }
                        
                        sketchCurve = sketch.sketchCurves.SketchFittedSplines.add(ptCol);
                    }
                    else if(geom instanceof adsk.core.InfiniteLine3D) {
                        var start = geom.origin;
                        var end = geom.origin;
                        var dir = geom.direction;
                        dir.scaleBy(10);
                        end.translateBy(dir);
                        sketchCurve = sketch.sketchCurves.sketchLines.addByTwoPoints(start, end);
                    }
                    
                    if(sketchCurve) {
                        sketchCurve.isConstruction = true;
                    }
                }
            }
        };
        
    };

    var createCommandDefinition = function() {
        var commandDefinitions = ui.commandDefinitions;

        // Be fault tolerant in case the command is already added...
        var cmdDef = commandDefinitions.itemById('IntersectionCMDDef');
        if (!cmdDef) {
            cmdDef = commandDefinitions.addButtonDefinition('IntersectionCMDDef', 
                    'Intersections', 
                    'Calculate the intersections of two selected entities',
                    './resources'); // relative resource file path is specified
        }
        return cmdDef;
    };
    
    var onCommandExecuted = function(args) {
        try {
            var command = args.command;
            
            var inputs = command.commandInputs;
            
            var input0 = inputs.item(0);
            var sel0 = input0.selection(0);

            var input1 = inputs.item(1);
            var sel1 = input1.selection(0);

            var intersections = new Intersections(sel0.entity, sel1.entity);
            intersections.Execute();
        }
        catch(e) {
            if (ui) {
                ui.messageBox('Failed to calculate the intersections :' + (e.description ? e.description : e));
            }
        }
    };
    
    var onCommandDestroy = function(args) { 
        adsk.terminate();
    };
    
    var onValidateInput = function(args) {
        try {
            var sels = ui.activeSelections;

            if(sels.count == 2) {
                args.areInputsValid = true;
            }
            else {
                args.areInputsValid = false;
            }
        }
        catch(e) {
            if (ui) {
                ui.messageBox('Failed :' + (e.description ? e.description : e));
            }
        }
    };
    
    var onCommandCreated = function(args) {
         try {
            var command = args.command;

            var validateEvent = command.validateInputs;
            validateEvent.add(onValidateInput);

            var commandEventExecuted = command.execute;
            commandEventExecuted.add(onCommandExecuted);
             
            var commandEventDestroy = command.destroy;
            commandEventDestroy.add(onCommandDestroy);

            var inputs = command.commandInputs;

            var i1 = inputs.addSelectionInput('entity', 'Entity One', 'Please select a curve, planear entity or a BRepBody, Component, Occurrence');

            i1.addSelectionFilter(adsk.core.SelectionCommandInput.Edges);
            i1.addSelectionFilter(adsk.core.SelectionCommandInput.PlanarFaces);
            i1.addSelectionFilter(adsk.core.SelectionCommandInput.SketchCurves);
            i1.addSelectionFilter(adsk.core.SelectionCommandInput.ConstructionLines);
            i1.addSelectionFilter(adsk.core.SelectionCommandInput.ConstructionPlanes);
            i1.addSelectionFilter(adsk.core.SelectionCommandInput.Bodies);
            i1.addSelectionFilter(adsk.core.SelectionCommandInput.Occurrences);
            i1.addSelectionFilter(adsk.core.SelectionCommandInput.RootComponents);

            var i2 = inputs.addSelectionInput('sectionentity', 'Entity Two', 'Please select a linear or planear entity');

            i2.addSelectionFilter(adsk.core.SelectionCommandInput.PlanarFaces);
            i2.addSelectionFilter(adsk.core.SelectionCommandInput.LinearEdges);
            i2.addSelectionFilter(adsk.core.SelectionCommandInput.SketchLines);
            i2.addSelectionFilter(adsk.core.SelectionCommandInput.ConstructionLines);
            i2.addSelectionFilter(adsk.core.SelectionCommandInput.ConstructionPlanes);
        }
        catch(e) {
            if (ui) {
                ui.messageBox('Failed :' + (e.description ? e.description : e));
            }
        }
    };
    
    try {
        if (adsk.debug === true) {
            /*jslint debug: true*/
            debugger;
            /*jslint debug: false*/
        }

        var cmdDef = createCommandDefinition();
        var cmdCreatedEvent = cmdDef.commandCreated;
        cmdCreatedEvent.add(onCommandCreated);
        
        cmdDef.execute();
    }
    catch(e) {
        if (ui) {
            ui.messageBox('Failed : ' + (e.description ? e.description : e));
        }
        
        adsk.terminate();
    }
}());
